File: PHP/flynn.php
<?php

require 'vendor/autoload.php'; // Assuming you are using Composer for dependencies
use React\Socket\Connector;
use React\EventLoop\Factory;
use React\Filesystem\Filesystem;
use React\Filesystem\FilesystemInterface;
use React\Filesystem\Node\FileInterface;
use React\Filesystem\Node\DirectoryInterface;
use React\Promise\Promise;

$cl = [
    'gr' => "\033[32m",
    'gb' => "\033[4m",
    'br' => "\033[34m",
    'st' => "\033[9m",
    'yl' => "\033[33m",
    'rt' => "\033[0m"
];

$socket = null;
$pingInterval = null;
$countdownInterval = null;
$potentialPoints = 0;
$countdown = "Calculating...";
$pointsTotal = 0;
$pointsToday = 0;
$reconnectAttempts = 0;
$maxReconnectAttempts = 5;
$maxReconnectInterval = 5 * 60; // 5 minutes in seconds
$CoderMarkPrinted = false;

function CoderMark() {
    global $CoderMarkPrinted, $cl;
    if (!$CoderMarkPrinted) {
        echo "
╭━━━╮╱╱╱╱╱╱╱╱╱╱╱╱╱╭━━━┳╮
┃╭━━╯╱╱╱╱╱╱╱╱╱╱╱╱╱┃╭━━┫┃{$cl['gr']}
┃╰━━┳╮╭┳━┳━━┳━━┳━╮┃╰━━┫┃╭╮╱╭┳━╮╭━╮
┃╭━━┫┃┃┃╭┫╭╮┃╭╮┃╭╮┫╭━━┫┃┃┃╱┃┃╭╮┫╭╮╮{$cl['br']}
┃┃╱╱┃╰╯┃┃┃╰╯┃╰╯┃┃┃┃┃╱╱┃╰┫╰━╯┃┃┃┃┃┃┃
╰╯╱╱╰━━┻╯╰━╮┣━━┻╯╰┻╯╱╱╰━┻━╮╭┻╯╰┻╯╰╯{$cl['rt']}
╱╱╱╱╱╱╱╱╱╱╱┃┃╱╱╱╱╱╱╱╱╱╱╱╭━╯┃
╱╱╱╱╱╱╱╱╱╱╱╰╯╱╱╱╱╱╱╱╱╱╱╱╰━━╯
\n{$cl['gb']}Teneo Node Cli {$cl['gr']}v1.1.0 {$cl['rt']}{$cl['gb']}{$cl['br']}dev_build{$cl['rt']}
        ";
        $CoderMarkPrinted = true;
    }
}

function readFileAsync($filename) {
    return new Promise(function ($resolve, $reject) use ($filename) {
        if (file_exists($filename)) {
            $data = file_get_contents($filename);
            $resolve(json_decode($data, true));
        } else {
            $resolve([]);
        }
    });
}

function writeFileAsync($filename, $data) {
    return new Promise(function ($resolve) use ($filename, $data) {
        file_put_contents($filename, json_encode($data));
        $resolve();
    });
}

function getLocalStorage() {
    return readFileAsync('localStorage.json');
}

function setLocalStorage($data) {
    return getLocalStorage()->then(function ($currentData) use ($data) {
        $newData = array_merge($currentData, $data);
        return writeFileAsync('localStorage.json', $newData);
    });
}

function getUserIdFromFile() {
    return readFileAsync('UserId.json')->then(function ($data) {
        return $data['userId'] ?? null;
    });
}

function setUserIdToFile($userId) {
    return writeFileAsync('UserId.json', ['userId' => $userId]);
}

function getAccountData() {
    return readFileAsync('DataAccount.json');
}

function setAccountData($email, $password, $access_token, $refresh_token, $personalCode) {
    $accountData = [
        'email' => $email,
        'password' => $password,
        'access_token' => $access_token,
        'refresh_token' => $refresh_token,
        'personalCode' => $personalCode
    ];
    return writeFileAsync('DataAccount.json', $accountData);
}

function getReconnectDelay($attempt) {
    global $maxReconnectInterval;
    $baseDelay = 5; // 5 seconds
    $additionalDelay = $attempt * 5; // Additional 5 seconds for each attempt
    return min($baseDelay + $additionalDelay, $maxReconnectInterval);
}

function connectWebSocket($userId) {
    global $socket, $reconnectAttempts, $countdown, $pointsTotal, $pointsToday;

    if ($socket) return;

    $version = "v0.2";
    $url = "wss://secure.ws.teneo.pro";
    $wsUrl = "{$url}/websocket?userId=" . urlencode($userId) . "&version=" . urlencode($version);
    $socket = new WebSocket($wsUrl);

    $socket->on('open', function () use (&$reconnectAttempts) {
        $reconnectAttempts = 0;
        $connectionTime = (new DateTime())->format(DateTime::ISO8601);
        setLocalStorage(['lastUpdated' => $connectionTime]);
        echo "WebSocket connected at " . $connectionTime . "\n";
        startPinging();
        startCountdownAndPoints();
    });

    $socket->on('message', function ($data) use (&$pointsTotal, &$pointsToday) {
        $data = json_decode($data, true);
        echo "Received message from WebSocket: " . json_encode($data) . "\n";
        if (isset($data['pointsTotal']) && isset($data['pointsToday'])) {
            $lastUpdated = (new DateTime())->format(DateTime::ISO8601);
            setLocalStorage([
                'lastUpdated' => $lastUpdated,
                'pointsTotal' => $data['pointsTotal'],
                'pointsToday' => $data['pointsToday'],
            ]);
            $pointsTotal = $data['pointsTotal'];
            $pointsToday = $data['pointsToday'];
        }
    });

    $socket->on('close', function ($event) {
        global $socket, $reconnectAttempts, $maxReconnectInterval;
        $socket = null;
        echo "WebSocket disconnected\n";
        stopPinging();
        if (!$event->wasClean) {
            $reconnectAttempts++;
            $delay = getReconnectDelay($reconnectAttempts);
            if ($delay < $maxReconnectInterval) {
                sleep($delay);
                reconnectWebSocket();
            } else {
                echo "Max reconnect interval reached. Giving up.\n";
            }
        }
    });

    $socket->on('error', function ($error) {
        echo "WebSocket error: " . $error . "\n";
    });
}

function disconnectWebSocket() {
    global $socket;
    if ($socket) {
        $socket->close();
        $socket = null;
        stopPinging();
    }
}

function startPinging() {
    global $pingInterval, $socket;
    stopPinging();
    $pingInterval = setInterval(function () use ($socket) {
        if ($socket && $socket->readyState === WebSocket::OPEN) {
            $socket->send(json_encode(['type' => 'PING']));
            setLocalStorage(['lastPingDate' => (new DateTime())->format(DateTime::ISO8601)]);
        }
    }, 10000);
}

function stopPinging() {
    global $pingInterval;
    if ($pingInterval) {
        clearInterval($pingInterval);
        $pingInterval = null;
    }
}

function startCountdownAndPoints() {
    global $countdownInterval;
    clearInterval($countdownInterval);
    updateCountdownAndPoints();
    $countdownInterval = setInterval('updateCountdownAndPoints', 1000);
}

function updateCountdownAndPoints() {
    global $countdown, $potentialPoints;
    getLocalStorage()->then(function ($localStorage) use (&$countdown, &$potentialPoints) {
        if (isset($localStorage['lastUpdated'])) {
            $nextHeartbeat = new DateTime($localStorage['lastUpdated']);
            $nextHeartbeat->modify('+15 minutes');
            $now = new DateTime();
            $diff = $nextHeartbeat->getTimestamp() - $now->getTimestamp();

            if ($diff > 0) {
                $minutes = floor($diff / 60);
                $seconds = $diff % 60;
                $countdown = "{$minutes}m {$seconds}s";

                $maxPoints = 25;
                $timeElapsed = $now->getTimestamp() - (new DateTime($localStorage['lastUpdated']))->getTimestamp();
                $timeElapsedMinutes = $timeElapsed / 60;
                $newPoints = min($maxPoints, ($timeElapsedMinutes / 15) * $maxPoints);
                $newPoints = round($newPoints, 2);

                if (mt_rand(0, 9) < 1) {
                    $bonus = mt_rand(0, 2);
                    $newPoints = min($maxPoints, $newPoints + $bonus);
                    $newPoints = round($newPoints, 2);
                }

                $potentialPoints = $newPoints;
            } else {
                $countdown = "Calculating...";
                $potentialPoints = 25;
            }
        } else {
            $countdown = "Calculating...";
            $potentialPoints = 0;
        }
        setLocalStorage(['potentialPoints' => $potentialPoints, 'countdown' => $countdown]);
    });
}

function getUserId() {
    global $cl;
    $loginUrl = "https://node-community-api.teneo.pro/auth/v1/token?grant_type=password";
    $authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";
    $apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";

    getAccountData()->then(function ($accountData) use ($loginUrl, $authorization, $apikey, $cl) {
        $email = $accountData['email'] ?? readline("\nEmail: ");
        $password = $accountData['password'] ?? readline($cl['rt'] . "Password: " . $cl['st'] . $cl['br'] . "");

        $client = new \GuzzleHttp\Client();
        $response = $client->post($loginUrl, [
            'json' => [
                'email' => $email,
                'password' => $password
            ],
            'headers' => [
                'Authorization' => $authorization,
                'apikey' => $apikey
            ]
        ]);

        $data = json_decode($response->getBody(), true);
        $access_token = $data['access_token'];
        $refresh_token = $data['refresh_token'];
        echo 'Access_Token: ' . $access_token . "\n";
        echo 'Refresh_Token: ' . $refresh_token . "\n";

        $AuthUserUrl = "https://node-community-api.teneo.pro/auth/v1/user";
        $AuthResponse = $client->get($AuthUserUrl, [
            'headers' => [
                'Authorization' => "Bearer {$access_token}",
                'apikey' => $apikey
            ]
        ]);

        $userId = json_decode($AuthResponse->getBody(), true)['id'];
        echo 'User ID: ' . $userId . "\n";

        $profileUrl = "https://node-community-api.teneo.pro/rest/v1/profiles?select=personal_code&id=eq.{$userId}";
        $profileResponse = $client->get($profileUrl, [
            'headers' => [
                'Authorization' => "Bearer {$access_token}",
                'apikey' => $apikey
            ]
        ]);
        $personalCode = json_decode($profileResponse->getBody(), true)[0]['personal_code'] ?? null;
        echo "Personal Code: " . $cl['rt'] . $personalCode . "\n";
        setUserIdToFile($userId);
        setAccountData($email, $password, $access_token, $refresh_token, $personalCode);
        startCountdownAndPoints();
        connectWebSocket($userId);
        echo $cl['gr'] . "Data has been saved in the DataAccount.json file...\n" . $cl['rt'];
        CoderMark();
    });
}

function reconnectWebSocket() {
    getUserIdFromFile()->then(function ($userId) {
        if ($userId) {
            connectWebSocket($userId);
        }
    });
}

function autoLogin() {
    getAccountData()->then(function ($accountData) {
        if (!empty($accountData['email']) && !empty($accountData['password'])) {
            getUserId();
            echo $cl['yl'] . "\nAutomatic Login has been Successfully Executed..\n" . $cl['rt'];
        }
    });
}

function main() {
    getLocalStorage()->then(function ($localStorageData) {
        getUserIdFromFile()->then(function ($userId) {
            global $cl;
            if (!$userId) {
                $option = readline("\nUser ID not found. Would you like to:\n" . $cl['gr'] . "\n1. Login to your account\n2. Enter User ID manually\n" . $cl['rt'] . "\nChoose an option: ");
                switch ($option) {
                    case '1':
                        getUserId();
                        break;
                    case '2':
                        $inputUserId = readline('Please enter your user ID: ');
                        setUserIdToFile($inputUserId);
                        startCountdownAndPoints();
                        connectWebSocket($inputUserId);
                        break;
                    default:
                        echo 'Invalid option. Exiting...' . "\n";
                        exit(0);
                }
            } else {
                $option = readline("\nMenu:\n" . $cl['gr'] . "\n1. Logout\n2. Start Running Node\n" . $cl['rt'] . "\nChoose an option: ");
                switch ($option) {
                    case '1':
                        unlink('UserId.json');
                        unlink('localStorage.json');
                        unlink('DataAccount.json');
                        echo $cl['yl'] . "\nLogged out successfully." . "\n";
                        exit(0);
                    case '2':
                        echo "\n";
                        CoderMark();
                        echo $cl['gr'] . "Initiates a connection to the node...\n" . $cl['rt'];
                        startCountdownAndPoints();
                        connectWebSocket($userId);
                        break;
                    default:
                        echo $cl['yl'] . "Invalid option. Exiting..." . "\n";
                        exit(0);
                }
            }
        });
    });
}

main();
setInterval('autoLogin', 1800);

File: RUBY/flynn.rb
require 'websocket-client-simple'
require 'json'
require 'fileutils'
require 'io/console'
require 'net/http'
require 'uri'

cl = { gr: "\e[32m", gb: "\e[4m", br: "\e[34m", st: "\e[9m", yl: "\e[33m", rt: "\e[0m" }

socket = nil
ping_interval = nil
countdown_interval = nil
potential_points = 0
countdown = "Calculating..."
points_total = 0
points_today = 0
reconnect_attempts = 0
max_reconnect_attempts = 5
max_reconnect_interval = 5 * 60 # 5 minutes in seconds
coder_mark_printed = false

def coder_mark
  return if coder_mark_printed

  puts <<~MARK
    ╭━━━╮╱╱╱╱╱╱╱╱╱╱╱╱╱╭━━━┳╮
    ┃╭━━╯╱╱╱╱╱╱╱╱╱╱╱╱╱┃╭━━┫┃#{cl[:gr]}
    ┃╰━━┳╮╭┳━┳━━┳━━┳━╮┃╰━━┫┃╭╮╱╭┳━╮╭━╮
    ┃╭━━┫┃┃┃╭┫╭╮┃╭╮┃╭╮┫╭━━┫┃┃┃╱┃┃╭╮┫╭╮╮#{cl[:br]}
    ┃┃╱╱┃╰╯┃┃┃╰╯┃╰╯┃┃┃┃┃╱╱┃╰┫╰━╯┃┃┃┃┃┃┃
    ╰╯╱╱╰━━┻╯╰━╮┣━━┻╯╰┻╯╱╱╰━┻━╮╭┻╯╰┻╯╰╯#{cl[:rt]}
    ╱╱╱╱╱╱╱╱╱╱╱┃┃╱╱╱╱╱╱╱╱╱╱╱╭━╯┃
    ╱╱╱╱╱╱╱╱╱╱╱╰╯╱╱╱╱╱╱╱╱╱╱╱╰━━╯
    \n#{cl[:gb]}Teneo Node Cli #{cl[:gr]}v1.1.0 #{cl[:rt]}#{cl[:gb]}#{cl[:br]}dev_build#{cl[:rt]}
  MARK

  coder_mark_printed = true
end

def read_file_async(file)
  JSON.parse(File.read(file))
rescue
  {}
end

def write_file_async(file, data)
  File.write(file, JSON.pretty_generate(data))
end

def get_local_storage
  read_file_async('localStorage.json')
end

def set_local_storage(data)
  current_data = get_local_storage
  new_data = current_data.merge(data)
  write_file_async('localStorage.json', new_data)
end

def get_user_id_from_file
  read_file_async('UserId.json')['userId']
rescue
  nil
end

def set_user_id_to_file(user_id)
  write_file_async('UserId.json', { userId: user_id })
end

def get_account_data
  read_file_async('DataAccount.json')
end

def set_account_data(email, password, access_token, refresh_token, personal_code)
  account_data = {
    email: email,
    password: password,
    access_token: access_token,
    refresh_token: refresh_token,
    personal_code: personal_code
  }
  write_file_async('DataAccount.json', account_data)
end

def get_reconnect_delay(attempt)
  base_delay = 5 # 5 seconds
  additional_delay = attempt * 5 # Additional 5 seconds for each attempt
  [base_delay + additional_delay, max_reconnect_interval].min
end

def connect_websocket(user_id)
  return if socket

  version = "v0.2"
  url = "wss://secure.ws.teneo.pro"
  ws_url = "#{url}/websocket?userId=#{URI.encode(user_id)}&version=#{URI.encode(version)}"
  socket = WebSocket::Client::Simple.connect(ws_url)

  socket.on :open do
    reconnect_attempts = 0
    connection_time = Time.now.iso8601
    set_local_storage(lastUpdated: connection_time)
    puts "WebSocket connected at #{connection_time}"
    start_pinging
    start_countdown_and_points
  end

  socket.on :message do |event|
    data = JSON.parse(event.data)
    puts "Received message from WebSocket: #{data}"
    if data['pointsTotal'] && data['pointsToday']
      last_updated = Time.now.iso8601
      set_local_storage(lastUpdated: last_updated, pointsTotal: data['pointsTotal'], pointsToday: data['pointsToday'])
      points_total = data['pointsTotal']
      points_today = data['pointsToday']
    end
  end

  socket.on :close do |event|
    socket = nil
    puts "WebSocket disconnected"
    stop_pinging
    unless event.was_clean
      reconnect_attempts += 1
      delay = get_reconnect_delay(reconnect_attempts)
      if delay < max_reconnect_interval
        sleep(delay)
        reconnect_websocket
      else
        puts "Max reconnect interval reached. Giving up."
      end
    end
  end

  socket.on :error do |error|
    puts "WebSocket error: #{error}"
  end
end

def disconnect_websocket
  if socket
    socket.close
    socket = nil
    stop_pinging
  end
end

def start_pinging
  stop_pinging
  ping_interval = Thread.new do
    loop do
      if socket && socket.ready_state == WebSocket::Client::Simple::OPEN
        socket.send(JSON.generate({ type: "PING" }))
        set_local_storage(lastPingDate: Time.now.iso8601)
      end
      sleep(10)
    end
  end
end

def stop_pinging
  if ping_interval
    ping_interval.kill
    ping_interval = nil
  end
end

Signal.trap("INT") do
  puts 'Received SIGINT. Stopping pinging...'
  stop_pinging
  disconnect_websocket
  exit(0)
end

def start_countdown_and_points
  countdown_interval = Thread.new do
    loop do
      update_countdown_and_points
      sleep(1)
    end
  end
end

def update_countdown_and_points
  last_updated = get_local_storage['lastUpdated']
  if last_updated
    next_heartbeat = Time.parse(last_updated) + 15 * 60
    now = Time.now
    diff = next_heartbeat - now

    if diff > 0
      minutes = (diff / 60).to_i
      seconds = (diff % 60).to_i
      countdown = "#{minutes}m #{seconds}s"

      max_points = 25
      time_elapsed = now - Time.parse(last_updated)
      time_elapsed_minutes = time_elapsed / 60
      new_points = [max_points, (time_elapsed_minutes / 15) * max_points].min
      new_points = new_points.round(2)

      if rand < 0.1
        bonus = rand * 2
        new_points = [max_points, new_points + bonus].min
        new_points = new_points.round(2)
      end

      potential_points = new_points
    else
      countdown = "Calculating..."
      potential_points = 25
    end
  else
    countdown = "Calculating..."
    potential_points = 0
  end
  set_local_storage(potentialPoints: potential_points, countdown: countdown)
end

def get_user_id
  login_url = "https://node-community-api.teneo.pro/auth/v1/token?grant_type=password"
  authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"
  apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"

  account_data = get_account_data
  email = account_data['email'] || (print "\nEmail: #{cl[:gr]}"; gets.chomp)
  password = account_data['password'] || (print "#{cl[:rt]}Password: #{cl[:st]}#{cl[:br]}"; gets.chomp)

  begin
    uri = URI.parse(login_url)
    response = Net::HTTP.post_form(uri, { email: email, password: password }, { 'Authorization' => authorization, 'apikey' => apikey })
    data = JSON.parse(response.body)

    access_token = data['access_token']
    refresh_token = data['refresh_token']
    puts "Access_Token: #{access_token}"
    puts "Refresh_Token: #{refresh_token}"

    auth_user_url = "https://node-community-api.teneo.pro/auth/v1/user"
    uri = URI.parse(auth_user_url)
    auth_response = Net::HTTP.get_response(uri, { 'Authorization' => "Bearer #{access_token}", 'apikey' => apikey })
    user_id = JSON.parse(auth_response.body)['id']
    puts "User ID: #{user_id}"

    profile_url = "https://node-community-api.teneo.pro/rest/v1/profiles?select=personal_code&id=eq.#{user_id}"
    uri = URI.parse(profile_url)
    profile_response = Net::HTTP.get_response(uri, { 'Authorization' => "Bearer #{access_token}", 'apikey' => apikey })
    personal_code = JSON.parse(profile_response.body)[0]['personal_code']
    puts "Personal Code: #{cl[:rt]}#{personal_code}"
    set_user_id_to_file(user_id)
    set_account_data(email, password, access_token, refresh_token, personal_code)
    start_countdown_and_points
    connect_websocket(user_id)
    system("clear")
    puts "#{cl[:gr]}Data has been saved in the DataAccount.json file...\n#{cl[:rt]}"
    coder_mark
  rescue => error
    puts "Error: #{error.message}"
  end
end

def reconnect_websocket
  user_id = get_user_id_from_file
  connect_websocket(user_id) if user_id
end

def auto_login
  account_data = get_account_data
  if account_data['email'] && account_data['password']
    get_user_id
    puts "#{cl[:yl]}\nAutomatic Login has been Successfully Executed..\n#{cl[:rt]}"
  end
end

def main
  local_storage_data = get_local_storage
  user_id = get_user_id_from_file

  if user_id.nil?
    print "\nUser ID not found. Would you like to:\n#{cl[:gr]}\n1. Login to your account\n2. Enter User ID manually\n#{cl[:rt]}\nChoose an option: "
    option = gets.chomp
    case option
    when '1'
      get_user_id
    when '2'
      print 'Please enter your user ID: '
      user_id = gets.chomp
      set_user_id_to_file(user_id)
      start_countdown_and_points
      connect_websocket(user_id)
    else
      puts 'Invalid option. Exiting...'
      exit(0)
    end
  else
    print "\nMenu:\n#{cl[:gr]}\n1. Logout\n2. Start Running Node\n#{cl[:rt]}\nChoose an option: "
    option = gets.chomp
    case option
    when '1'
      FileUtils.rm_f('UserId.json')
      FileUtils.rm_f('localStorage.json')
      FileUtils.rm_f('DataAccount.json')
      puts "#{cl[:yl]}\nLogged out successfully."
      exit(0)
    when '2'
      system("clear")
      coder_mark
      puts "#{cl[:gr]}Initiates a connection to the node...\n#{cl[:rt]}"
      start_countdown_and_points
      connect_websocket(user_id)
    else
      puts "#{cl[:yl]}Invalid option. Exiting..."
      exit(0)
    end
  end
end

main
Thread.new { loop { auto_login; sleep(1800) } }

